databricks/koalas/tests/test_indexes.py

```diff
@@ -1390,6 +1390,80 @@ def test_hasnans(self):
         kser = ks.from_pandas(pser)
         self.assert_eq(pser.hasnans, kser.hasnans)
 
+    def test_intersection(self):
+        pidx = pd.Index([1, 2, 3, 4])
+        kidx = ks.from_pandas(pidx)
+
+        # other = Index
+        pidx_other = pd.Index([3, 4, 5, 6])
+        kidx_other = ks.from_pandas(pidx_other)
+        self.assert_eq(pidx.intersection(pidx_other), kidx.intersection(kidx_other).sort_values())
+        self.assert_eq(
+            (pidx + 1).intersection(pidx_other), (kidx + 1).intersection(kidx_other).sort_values()
+        )
+
+        # other = MultiIndex
+        pmidx = pd.MultiIndex.from_tuples([("a", "x"), ("b", "y"), ("c", "z")])
+        kmidx = ks.from_pandas(pmidx)
+        self.assert_eq(
+            pidx.intersection(pmidx), kidx.intersection(kmidx).sort_values(), almost=True
+        )
+        self.assert_eq(
+            (pidx + 1).intersection(pmidx),
+            (kidx + 1).intersection(kmidx).sort_values(),
+            almost=True,
+        )
+
+        # other = Series
+        pser = pd.Series([3, 4, 5, 6])
+        kser = ks.from_pandas(pser)
+        self.assert_eq(pidx.intersection(pser), kidx.intersection(kser).sort_values())
+        self.assert_eq((pidx + 1).intersection(pser), (kidx + 1).intersection(kser).sort_values())
+
+        # other = list
+        other = [3, 4, 5, 6]
+        self.assert_eq(pidx.intersection(other), kidx.intersection(other).sort_values())
+        self.assert_eq((pidx + 1).intersection(other), (kidx + 1).intersection(other).sort_values())
+
+        # other = tuple
+        other = (3, 4, 5, 6)
+        self.assert_eq(pidx.intersection(other), kidx.intersection(other).sort_values())
+        self.assert_eq((pidx + 1).intersection(other), (kidx + 1).intersection(other).sort_values())
+
+        # other = dict
+        other = {3: None, 4: None, 5: None, 6: None}
+        self.assert_eq(pidx.intersection(other), kidx.intersection(other).sort_values())
+        self.assert_eq((pidx + 1).intersection(other), (kidx + 1).intersection(other).sort_values())
+
+        # MultiIndex / other = Index
+        self.assert_eq(
+            pmidx.intersection(pidx), kmidx.intersection(kidx).sort_values(), almost=True
+        )
+
+        # MultiIndex / other = MultiIndex
+        pmidx_other = pd.MultiIndex.from_tuples([("c", "z"), ("d", "w")])
+        kmidx_other = ks.from_pandas(pmidx_other)
+        self.assert_eq(
+            pmidx.intersection(pmidx_other), kmidx.intersection(kmidx_other).sort_values()
+        )
+
+        # MultiIndex / other = list
+        other = [("c", "z"), ("d", "w")]
+        self.assert_eq(pmidx.intersection(other), kmidx.intersection(other).sort_values())
+
+        # MultiIndex / other = tuple
```

Comment:
How does it work if `MultiIndex / other = Series`?

Reply:
Excluded because Koalas cannot construct `Series` with `tuple` type.

```python
>>> ks.Series([("c", "z"), ("d", "w")])
Traceback (most recent call last):
...
pyarrow.lib.ArrowInvalid: Could not convert ('c', 'z') with type tuple: did not recognize Python value type when inferring an Arrow data type
```